### PR TITLE
refactor(gcp-link-account): Refactor GCP integration resource for readability

### PR DESCRIPTION
**Pull request by rpaliwal1997**
[Jira Ticket](https://new-relic.atlassian.net/browse/NR-347137)

**Refactor GCP Integration Resource**

**Description:** This pull request includes a refactor of the GCP integration resource in the [resource_newrelic_cloud_gcp_integrations.go](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) file to improve code readability and maintainability. 

**The changes include:**

- Refactored the [resourceNewrelicCloudGcpIntegrations](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function to use a new [generateGcpIntegrationSchema](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function for schema generation.

- Reduced cyclomatic complexity in the flattenCloudGcpLinkedAccount function.

**Changes:**

- Added [generateGcpIntegrationSchema](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function to encapsulate schema generation logic.

- Refactored flattenCloudGcpLinkedAccount to handle different GCP integrations more cleanly.

- Improved code structure and organization for better readability.

**Impact:** These changes do not alter the functionality of the GCP integration resource but make the codebase easier to understand and maintain.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

Please delete options that are not relevant.

- [ ] My commit message follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] My code is formatted to [Go standards](https://go.dev/blog/gofmt)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes. Go [here](https://github.com/newrelic/terraform-provider-newrelic/blob/main/CONTRIBUTING.md#testing) for instructions on running tests locally.

## How to test this change?

Please describe how to test your changes. Include any relevant steps in the UI, HCL file(s), commands, etc

- Step 1
- Step 2
- etc

### CODE REVIEW

File: newrelic/resource_newrelic_cloud_gcp_integrations.go [hunk 1]
```diff
@@ -19,201 +19,206 @@ func resourceNewrelicCloudGcpIntegrations() *schema.Resource {
...
-				Optional:    true,
-				Elem:        cloudGcpIntegrationsVpcAccessSchemaElem(),
-				MaxItems:    1,
-			},
+		Schema: generateGcpIntegrationSchema(),
+	}
+}
+
+func generateGcpIntegrationSchema() map[string]*schema.Schema {
+	baseSchema := cloudGcpIntegrationSchemaBase()
```
Comment (pranav-new-relic) [inline newrelic/resource_newrelic_cloud_gcp_integrations.go:27]:
```suggestion
	gcp_integrations_base_schema := cloudGcpIntegrationSchemaBase()
```

(just trying to correct the case, as generally used in Golang :) please rename this variable)

File: newrelic/resource_newrelic_cloud_gcp_integrations.go [hunk 2]
```diff
@@ -229,232 +234,17 @@ func cloudGcpIntegrationSchemaBase() map[string]*schema.Schema {
...
-		Schema: s,
-	}
-}
-
-// function to add schema for gcp load balancing
-func cloudGcpIntegrationsLoadBalancingSchemaElem() *schema.Resource {
-	s := cloudGcpIntegrationSchemaBase()
-	return &schema.Resource{
-		Schema: s,
+func addFetchTagsToSchema(baseSchema map[string]*schema.Schema) map[string]*schema.Schema {
```
Comment (pranav-new-relic) [inline newrelic/resource_newrelic_cloud_gcp_integrations.go:237]: @rpaliwal1997 amazing work on writing a common function containing the schema for most integrations :) this has helped us get rid of 200 lines of code.

I have a suggestion about this function. The purpose of this function is only to be called for the integrations which have an extra `fetch_tags` argument (alongside `metrics_polling_interval`), i.e. `big_query`, `pub_sub`, `spanner` and `storage`; so instead of having this function `addFetchTagsToSchema`, I would suggest using a similar approach using a helper function called `mergeSchemas()` we already have in the provider.

https://github.com/newrelic/terraform-provider-newrelic/blob/25894e484177876891d4407de299bd4692f20e6e/newrelic/helpers.go#L125-L133

We already have a `cloudGcpIntegrationsSchemaBase()` which exports `metrics_polling_interval`; we can have a `cloudGcpIntegrationsCustomSchema()` which exports `fetch_tags` similar to `cloudGcpIntegrationsSchemaBase()`. Using a new function in this file (say `cloudGcpIntegrationsAddCustomSchemaToBase()`), we can use `mergeSchemas()` to combine schemas from both functions, and the output of this function can directly be referenced in the schema of `pub_sub`, `spanner`, `storage`, `big_query`. I wanted to suggest this as we can use this format to keep code consistent across the provider.

A very recent example of using this helper function is `cloudAzureIntegrationMergeResourceGroupsElem()` added/modified recently by Aashirwad. Search for this function on the main branch, and you'll get an idea of what I mean to say, as the usage of this function is pretty much similar.

File: newrelic/resource_newrelic_cloud_gcp_integrations.go [hunk 3]
```diff
@@ -1218,294 +1008,124 @@ func flattenCloudGcpLinkedAccount(d *schema.ResourceData, linkedAccount *cloud.C
...
 		case *cloud.CloudGcpVpcaccessIntegration:
-			_ = d.Set("vpc_access", flattenCloudGcpVpcAccessIntegration(t))
+			_ = d.Set("vpc_access", flattenCloudGcpCommonIntegration(t))
 		}
 	}
 }
 
-// flatten function to set(store) outputs from the terraform apply
-func flattenCloudGcpAppEngineIntegration(in *cloud.CloudGcpAppengineIntegration) []interface{} {
+func flattenCloudGcpCommonIntegration(in interface{}) []interface{} {
```
Comment (pranav-new-relic) [inline newrelic/resource_newrelic_cloud_gcp_integrations.go:1066]: @rpaliwal1997 I have been deliberating/trying/experimenting to get rid of this function/simplifying the switch cases (while I fully know why you needed to add this :)) as we now have two functions instead of one with the exact same cases (of course, this is helping us remove so many other functions, but it doesn't really seem logical that we have two functions with the same cases).

However, from all of my experimentation in the last bunch of hours, I realize this is needed if we want to get rid of rest of the functions, because Go is such a rigid typed language; it all comes down to the fact that we cannot have a generic case written to match all of these types to get `t.MetricsPollingInterval` because all of these `GcpCloudIntegration` types are implementations of the `CloudIntegrationInterface` in Golang, but they don't have a custom implementation per class of `GetMetricsPollingInterval` to return the metrics polling interval (which would have been useful for our condition, but that's just how Tutone generates this from NerdGraph). Ideally the only other way to make this possible would be to write some custom interfaces of our own in the Go Client, but that is a bad solution.

In summary; I just wanted to share all this for general awareness :) since it looks like there's no other way, we can probably try to explore if there's still some other approach we're missing in the future, but for now, since the lint failure you're seeing on the PR is because of the cyclomatic complexity of this function, please add a `// nolint:gocyclo` comment on top of the function similar to what you will find above `flattenCloudGcpLinkedAccount`. We'll get the PR merged with this for now and then think of a probable, better solution.

**Review by pranav-new-relic — DISMISSED** [review]

**Comment by pranav-new-relic** [comment]
@rpaliwal1997 I've asked for some changes in three comments; please take a look, fix them and then we should be good to go with a merge (but no release until the code freeze is done). Just make sure you get an extra pair of eyes before the merge 👍

File: newrelic/resource_newrelic_cloud_gcp_integrations.go (hunk 1 above)
Comment (rpaliwal1997) [reply newrelic/resource_newrelic_cloud_gcp_integrations.go]: Thanks for the suggestion, got it, I renamed the variable(gcpIntegrationsBaseSchema)

File: newrelic/resource_newrelic_cloud_gcp_integrations.go (hunk 2 above)
Comment (rpaliwal1997) [reply newrelic/resource_newrelic_cloud_gcp_integrations.go]: @pranav-new-relic I got it, I made the suggested the changes. Thank you

File: newrelic/resource_newrelic_cloud_gcp_integrations.go (hunk 3 above)
Comment (rpaliwal1997) [reply newrelic/resource_newrelic_cloud_gcp_integrations.go:1077]: @pranav-new-relic After adding // nolint:gocyclo in the function **flattenCloudGcpCommonIntegration** , GoLint tests are passed.

**Review by pranav-new-relic — APPROVED** [review]
👍

**Review by vagrawal-newrelic — COMMENTED** [review]
approving
